tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15541,46 +15541,48 @@ void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, llvm::StringRef StageName,
-      for (FunctionDecl::param_iterator I = FD->param_begin(),
-                                        E = FD->param_end();
-           I != E; ++I) {
-        QualType ParamType = (*I)->getType().getCanonicalType();
-
-        // Find parameter that is the node input record
-        if (hlsl::IsHLSLNodeInputType(ParamType)) {
-          // Node input records are template types
-          if (const RecordType *NodeInputRT = dyn_cast<RecordType>(ParamType)) {
-            if (const ClassTemplateSpecializationDecl *templateDecl =
-                    dyn_cast<ClassTemplateSpecializationDecl>(
-                        NodeInputRT->getDecl())) {
-
-              // Get the input record struct
-              auto &TemplateArgs = templateDecl->getTemplateArgs();
-              DXASSERT_NOMSG(TemplateArgs.size() >= 1);
-              QualType Arg0Type = TemplateArgs.get(0).getAsType();
-              const RecordType *NodeInputStructType =
-                  Arg0Type->getAsStructureType();
-              if (nullptr != NodeInputStructType) {
-                CXXRecordDecl *NodeInputStructDecl =
-                    dyn_cast<CXXRecordDecl>(NodeInputStructType->getDecl());
-                if (nullptr != NodeInputStructDecl) {
-                  // Make sure there is exactly one SV_DispatchGrid semantics
-                  // and it has correct type.
-                  DiagnoseDispatchGridSemantics(S, NodeInputStructDecl, Found);
-                }
+    // Diagnose dispatch grid semantics.
+    bool Found = false;
+    for (FunctionDecl::param_iterator I = FD->param_begin(),
+                                      E = FD->param_end();
+         I != E; ++I) {
+      QualType ParamType = (*I)->getType().getCanonicalType();
+
+      // Find parameter that is the node input record
+      if (hlsl::IsHLSLNodeType(ParamType)) {
+        // Node records are template types
+        if (const RecordType *NodeRT = dyn_cast<RecordType>(ParamType)) {
+          if (const ClassTemplateSpecializationDecl *templateDecl =
+                  dyn_cast<ClassTemplateSpecializationDecl>(
+                      NodeRT->getDecl())) {
+
+            // Get the record struct
+            auto &TemplateArgs = templateDecl->getTemplateArgs();
+            DXASSERT_NOMSG(TemplateArgs.size() >= 1);
+            QualType Arg0Type = TemplateArgs.get(0).getAsType();
+            const RecordType *NodeStructType = Arg0Type->getAsStructureType();
+            if (nullptr != NodeStructType) {
+              CXXRecordDecl *NodeStructDecl =
+                  dyn_cast<CXXRecordDecl>(NodeStructType->getDecl());
+              if (nullptr != NodeStructDecl) {
+                bool OutputFound = false;
+                // Make sure there is exactly one SV_DispatchGrid semantics
+                // and it has correct type.
+                DiagnoseDispatchGridSemantics(
+                    S, NodeStructDecl,
+                    hlsl::IsHLSLNodeOutputType(ParamType) ? OutputFound
```

Comment:
In this code, within this for loop, Found is set to false and never re-assigned, and OutputFound is set to false and never re-assigned. Then, they're both used at the bottom in DiagnoseDispatchGridSemantics(). I think both variables can be removed, and we can substitute `false` instead? We may need a refactor or to re-inspect the logic here, because I don't think these variables are being used correctly.

Reply:
DiagnoseDispatchGridSemantics will update Found for input node types.